packages/eslint-plugin/src/rules/no-extraneous-return-types.ts

```diff
@@ -0,0 +1,311 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import type * as ts from 'typescript';
+
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+
+import { createRule, forEachReturnStatement, getParserServices } from '../util';
+
+export default createRule({
+  name: 'no-extraneous-return-types',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow extraneous union types in return type annotations',
+      requiresTypeChecking: true,
+    },
+    fixable: 'code',
+    messages: {
+      unusedReturnTypes:
+        "Unused or redundant union type '{{type}}' in return type.",
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    function getRelevantReturnTypes(
+      node: TSESTree.TypeNode,
+      returnTypes: ts.Type[],
+    ): ts.Type[] | null {
+      const type = services.getTypeAtLocation(node);
+
+      const relevantReturnTypes = returnTypes.filter(returnType =>
+        checker.isTypeAssignableTo(returnType, type),
+      );
+
+      if (relevantReturnTypes.length === 0) {
+        context.report({
+          node,
+          messageId: 'unusedReturnTypes',
+          data: {
+            type: checker.typeToString(type),
+          },
+        });
+
+        return null;
+      }
+
+      return relevantReturnTypes;
+    }
+
+    function checkReturnType(
+      node: TSESTree.TypeNode,
+      returnTypes: ts.Type[],
+      allowUnusedVoidOrUndefinedValues = false,
+    ): void {
```

Comment:
[Style] This is totally optional and if you don't like it I won't be upset at all 😄, but: I found this function to be a lot to read through. It's essentially a set of individual cases to check: tuple types, promise type argument, array type argument, miscellaneous. It might be clearer to split it up?

```ts
function checkReturnType(...) {
  const tupleElementTypes = getTypeArgumentIfTupleNode(node);
  if (tupleElementTypes) {
    checkTupleElementReturnTypes(...);
    return;
  }

  const promiseTypeArgument = getTypeArgumentIfPromiseNode(node);
  if (promiseTypeArgument) {
    checkPromiseTypeArgumentReturnTypes(...);
    return;
  }

  // (similar const/if/check/return for array type arguments)

  // (similar if/check/return for union types)

  // ...miscellaneous logic
}
``` 

If you don't like this, not a problem at all - just a personal preference of mine.